lessons/PowerShell/managed-identity-powershell.md

```diff
@@ -0,0 +1,240 @@
+
+* we don’t need to adjust to ever changing UI in Azure portal
+* we can benefit from source control
+* we can collaborate with others
+
+### Steps
+
+1. Install the Core Tools package with `npm install -g azure-functions-core-tools@3 --unsafe-perm true`
+2. Install the [Azure Functions extension for VS Code](https://marketplace.visualstudio.com/items?itemName=ms-azuretools.vscode-azurefunctions)
+3. Select **New Project**
+4. Select a folder for your project
+5. Select a language – I will use PowerShell
+6. Select **HTTP trigger** as a template
+7. Type in a better name like `GetGraphToken`
+8. Select Authorization level **Function**
+9. Select how you want to open your project – I prefer **Add to workspace**
+10. Open `run.ps1`
+11. Replace the default code by this: 
+
+```
+using namespace System.Net
+
+# Input bindings are passed in via param block
+param($Request, $TriggerMetadata)
+
+# Write to the Azure Functions log stream.
+Write-Host "PowerShell HTTP trigger function processed a request."
+
+# Interact with query parameters or the body of the request
+.$Scope = $Request.Query.Scope
+if (-not $Scope) {
+    $Scope = $Request.Body.Scope
+}
+#If parameter "Scope" has not been provided, we assume that graph.microsoft.com is the target resource
+If (!$Scope) {
+    $Scope = "https://graph.microsoft.com/"
+}
+
+$tokenAuthUri = $env:IDENTITY_ENDPOINT + "?resource=$Scope&api-version=2019-08-01"
+$response = Invoke-RestMethod -Method Get -Headers @{"X-IDENTITY-HEADER"="$env:IDENTITY_HEADER"} -Uri $tokenAuthUri -UseBasicParsing
+
+$accessToken = $response.access_token
+
+#Invoke REST call to Graph API
+$uri = 'https://graph.microsoft.com/v1.0/groups'
+$authHeader = @{    
+'Content-Type'='application/json'
+'Authorization'='Bearer ' +  $accessToken
+}
+
+$result = (Invoke-RestMethod -Uri $uri -Headers $authHeader -Method Get -ResponseHeadersVariable RES).value
+
+If ($result) {
+    $body = $result
+    $StatusCode = '200'
+}
+Else {
+    $body = $RES
+    $StatusCode = '400'}
+
+# Associate values to output bindings by calling 'Push-OutputBinding'
+Push-OutputBinding -Name Response -Value ([HttpResponseContext]@{
+    StatusCode = $StatusCode
+    Body = $body
+})
+```
+> 📝 **Tip** - Watch out – the Graph API will this way return up to 100 groups – Please adjust with query parameters as needed like `https://graph.microsoft.com/v1.0/groups?$top=42` or use pagination, which is described here: [Paging Microsoft Graph data in your app](https://docs.microsoft.com/en-us/graph/paging)
+
+Take a moment to understand what the code does: 
+
+* log that a request was received
+* define the scope (if not stated, it’s Microsoft Graph)
+* obtain the token from the environment variables IDENTITY_ENDPOINT and IDENTITY_HEADER (more about that in the next step!)
+* pass that token in the header of the REST call towards the group endpoint of Microsoft Graph
+* get the status code (hopefully 200) 🤞
+
+## 2. Create Azure resources
```

Comment:
This section has a huge overlap with the existing Deployment lesson. That lesson was written with a .NET function in mind. I'll either update that lesson or add a new one that is specific to PowerShell, so we can refer to that lesson and not duplicate the content here.

Reply:
❓ so I will just add a "For deployment see LINK"  and delete this section?